src/Microsoft.Extensions.Configuration.AzureAppConfiguration/DomainVerifier.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright(c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Security;
+using System.Net.Sockets;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Configuration.AzureAppConfiguration
+{
+    /// <summary>
+    /// Verify the domain name of the endpoint matches the certificate.
+    /// </summary>
+    internal static class DomainVerifier
+    {
+        private const int TlsPort = 443;
+        private const string SubjectAltNameOid = "2.5.29.17";
+        private const string CommonName = "CN=";
+        private const string MultiDomainWildcard = "*.";
+
+        public static async Task<string> GetValidDomain(Uri originEndpoint, string srvHostName)
+        {
+            if (string.IsNullOrEmpty(srvHostName))
+            {
+                return string.Empty;
+            }
+
+            IEnumerable<string> validDomains = await GetValidDomainsFromTlsCert(srvHostName).ConfigureAwait(false);
+            
+            foreach (var domain in validDomains)
+            {
+                if (originEndpoint.DnsSafeHost.EndsWith(domain, StringComparison.OrdinalIgnoreCase))
+                {
+                    return domain;
+                }
+            }
+
+            return string.Empty;
+        }
+
+        private static async Task<List<string>> GetValidDomainsFromTlsCert(string endpoint)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(endpoint));
+
+            var validDomains = new List<string>();
+
+            X509Certificate2 cert = null;
+
+            using (var client = new TcpClient(endpoint, TlsPort))
+            using (var sslStream = new SslStream(client.GetStream(), leaveInnerStreamOpen: false))
+            {
+
+                // Initiate the connection, so it will download the server certificate
+                await sslStream.AuthenticateAsClientAsync(endpoint).ConfigureAwait(false);
+
+                var serverCertificate = sslStream.RemoteCertificate;

```

Comment:
This will throw for untrusted cert right?

Reply:
I think not, this doesn't verify whether it's a trusted, it won't verify the certificate chain, the expiration time, the issuer or the root cert is trusted or not. And seems doing such verification is overkill here because we just aim to verify the domain of srv endpoint, not really need to check the legality of that certificate. We just leverage the cert to ensure the acquired SRV endpoints are in the same domain as the original endpoint would be fine.

Reply:
No way ! We shouldn't be using any info from a certificate we haven't validated.

Reply:
In that case, someone controlling a bad endpoint could mint a certificate with whatever they want.

Reply:
Then we should validate the certificate, what kind of validation should be done against the certificate? This is what we can get which can help to do the validation
https://learn.microsoft.com/en-us/dotnet/api/system.net.security.sslpolicyerrors?view=net-7.0

Reply:
Can we make an anonymous http call to the endpoint? I think http call has built in validation.

Reply:
https://learn.microsoft.com/en-us/dotnet/api/system.net.security.remotecertificatevalidationcallback?view=net-7.0
Such Example

Reply:
It does throw for untrusted cert as expected.